eli5/keras.py

```diff
@@ -0,0 +1,465 @@
+
+    
+    Returns
+    -------
+    (heatmap, predicted_idx, score) : (numpy.ndarray, int, float)
+        A Grad-CAM localization map,
+        the predicted class ID, and
+        the score (i.e. probability) for that class.
+
+
+    Notes
+    -----
+    We currently make two assumptions in this implementation
+        * We are dealing with images as our input to ``estimator``.
+        * We are doing a classification. Our ``estimator``'s output is a class scores vector.
+
+    Credits
+        * Jacob Gildenblat for "https://github.com/jacobgil/keras-grad-cam".
+        * Author of "https://github.com/PowerOfCreation/keras-grad-cam" for fixes to Jacob's implementation.
+        * Kotikalapudi, Raghavendra and contributors for "https://github.com/raghakot/keras-vis".
+    """
+    # Get required terms
+    weights, activations, grads, predicted_idx, score = grad_cam_backend(estimator, doc, targets, activation_layer)
+
+    # TODO: might want to replace np operations with keras backend operations
+
+    # Perform a weighted linear combination
+    spatial_shape = activations.shape[:2]
+    lmap = np.zeros(spatial_shape, dtype=np.float32)
+    for i, w in enumerate(weights):
+        # weight * single activation map
+        # add to the entire map (linear combination), NOT pixel by pixel
+        lmap += w * activations[..., i]
+
+    lmap = np.maximum(lmap, 0) # ReLU
+
+    # normalize lmap to [0, 1] ndarray
+    # add eps to avoid division by zero in case lmap is 0's
+    # this also means that lmap max will be slightly less than the 'true' max
+    lmap = lmap / (np.max(lmap)+K.epsilon())
+    return lmap, predicted_idx, score
+
+
+def grad_cam_backend(estimator, # type: Model
+    doc, # type: np.ndarray
+    targets, # type: Optional[List[int]]
+    activation_layer # type: Layer
+    ):
+    # type: (...) -> Tuple[np.ndarray, np.ndarray, np.ndarray, int, float]
+    """
+    Compute the terms required by the Grad-CAM formula and its by-products.
+    
+    See :func:`eli5.keras.grad_cam` for description of the 
+    ``estimator``, ``doc``, ``targets``, and ``activation_layer``
+    parameters.
+
+    Returns
+    -------
+    (weights, activations, gradients, predicted_idx, score) : (numpy.ndarray, ..., int, float)
+        Values of variables.
+    """
+    output = estimator.output
+    predicted_idx = get_target_prediction(targets, output)
+    score = K.gather(output[0,:], predicted_idx) # access value by index
+
+    # output of target layer, i.e. activation maps of a convolutional layer
+    activation_output = activation_layer.output 
+
+    grads = K.gradients(score, [activation_output])
+    # FIXME: this might have issues
+    # See https://github.com/jacobgil/keras-grad-cam/issues/17
+    # a fix is the following piece of code:
+    # grads = [grad if grad is not None else K.zeros_like(var) 
+    #         for (var, grad) in zip(xs, grads)]
+    grads = grads[0]
```

Comment:
When can we have more gradients? Do we want to use them? Maybe silently discarding them is ok, but it would be good to add an explanation here.

Reply:
Our current `grads` definition (`grads = K.gradients(score, [activation_output])`) has only one variable for "ys" (`activation_output`), so there will be only one tensor in the `grads` singleton list. (`K.function` and other operations require tensors).

(Added comment about this)

Reply:
Thanks @teabolt 👍 In this case, I suggest adding an assert or writing something like ``grads, = grads``, so that if there are more gradients than we expect, we'd know about this.

Reply:
Great solution @lopuhin thanks.

https://github.com/TeamHG-Memex/eli5/pull/315/commits/684465b048f1a9f7169d02a241a842246b8cd8c7